File: Algorithms/Searching/minmax.py
def minmax(data, low, high):
    if low == high:
        return {"min" : low, "max" : low}

    mid = (low + high) // 2
    m1 = minmax(data, low, mid)
    m2 = minmax(data, mid + 1, high)

    return {"min" : m1["min"] if data[m1["min"]] < data[m2["min"]] else m2["min"],
            "max" : m1["max"] if data[m1["max"]] > data[m2["max"]] else m2["max"]}

if __name__ == "__main__":
    data = [102, 21, 4, 31, 12, 32]
    print(minmax(data, 0, len(data) - 1))

File: Algorithms/Searching/minmax.cpp
#include <iostream>
#include <vector>

using namespace std;

void print(vector<int> data) {
  for (auto e : data)
    cout << e << "\n";
  cout << endl;
}

// Min Max
map<string, int> minmax(vector<int> &data, int low, int high) {
  map<string, int> m;

  if (low == high) {
    m["Min"] = data[low];
    m["Max"] = data[low];
    return m;
  }

  int mid = (low + high) / 2;

  map<string, int> m1 = minmax(data, low, mid);
  map<string, int> m2 = minmax(data, mid + 1, high);

  m["Min"] = (m1["Min"] < m2["Min"]) ? m1["Min"] : m2["Min"];
  m["Max"] = (m1["Max"] > m2["Max"]) ? m1["Max"] : m2["Max"];

  return m;
}

map<string, int> minmax(vector<int> data) {
  map<string, int> m;

  if (data.size() <= 0) {
    m["Min"] = -1;
    m["Max"] = -1;
    cout << "Inside if" << endl;
    return m;
  }

  return minmax(data, 0, (int)data.size() - 1);
}

int main() {
  vector<int> data = generateRandomIntVector(10);

  cout << "Data: \n";
  print(data);

  cout << "\n==== Min Max =====\n";

  map<string, int> result = minmax(data);

  for (auto r : result)
    cout << " => " << r.first << " : " << r.second << endl;
}
